File: src/lib/prediction-engine.ts
import { prisma } from "./prisma";

export type PredictionInput = {
  percentile: number;
  marks?: number;
  category: string;
  gender?: string;
  homeUniversity?: string;
  tfws?: boolean;
  minority?: boolean;
  preferredBranches?: string[];
  preferredColleges?: string[];
  city?: string;
  year?: number;
  capRound?: number;
};

export type PredictionResult = {
  collegeId: string;
  collegeName: string;
  collegeSlug: string;
  branchId: string;
  branchName: string;
  category: string;
  closingPercentile: number;
  openingPercentile: number;
  closingRank: number;
  openingRank: number;
  year: number;
  capRound: number;
  level: "dream" | "moderate" | "safe" | "no-chance";
  score: number;
};

// Mapping of parent cities to their known suburbs/areas
// When a user selects a parent city, also search its suburbs
const CITY_SUBURBS: Record<string, string[]> = {
  "Navi Mumbai": ["Panvel", "Vashi", "Nerul", "Airoli", "Ghansoli", "Belapur", "Kharghar", "Kamothe", "Kalamboli", "New Panvel"],
  "Mumbai": ["Andheri", "Bhayander", "Bhiwandi", "Kandivali", "Matunga", "Boisar", "Palghar"],
  "Pune": ["Pimpri", "Chinchwad", "Haveli", "Pisoli", "Ravet", "Sasewadi", "Talegaon", "Wagholi", "Avasari Khurd"],
  "Sangli": ["Miraj"],
  "Amravati": ["Badnera", "Shegaon"],
  "Nashik": ["Nepti", "Nashik", "Nadurbar"],
  "Kalyan": ["Dombivli", "Ulhasnagar"],
  "Kolhapur": ["Ichalkaranji", "Panhala"],
  "Solapur": ["Barshi"],
  "Nagpur": ["Ramtek", "Wardha"],
};

// Build reverse lookup: suburb → parent city
const SUBURB_TO_PARENT: Record<string, string> = {};
for (const [parent, suburbs] of Object.entries(CITY_SUBURBS)) {
  for (const suburb of suburbs) {
    SUBURB_TO_PARENT[suburb.toLowerCase()] = parent.toLowerCase();
  }
}

// Cache for the category variants mapping
let categoryVariantsCache: Promise<Map<string, string[]>> | null = null;

/**
 * Build a mapping from any category to all its seat-level variants.
 * 
 * Categories in the DB follow pattern: {prefix}{base}{seat_level}
 * where seat_level ∈ {S (State), H (Home University), O (Other)}
 * 
 * E.g., "GOPENS" → ["GOPENS", "GOPENH", "GOPENO"]
 *       "GSCS"  → ["GSCS", "GSCH", "GSCO"]
 *       "TFWS"  → ["TFWS"] (no seat-level variants)
 */
async function buildCategoryVariants(): Promise<Map<string, string[]>> {
  // Fetch all categories (avoid Prisma distinct issues, deduplicate in JS)
  const rows = await prisma.cutoff.findMany({
    select: { category: true },
  });

  const seen = new Set<string>();
  const allCategories = rows
    .map((r) => r.category)
    .filter((cat) => {
      if (seen.has(cat)) return false;
      seen.add(cat);
      return true;
    });
  
  const groups = new Map<string, string[]>();

  // Group by base category (strip trailing S/H/O if applicable)
  for (const cat of allCategories) {
    const lastChar = cat.slice(-1);
    const base =
      lastChar === "S" || lastChar === "H" || lastChar === "O"
        ? cat.slice(0, -1)
        : cat;
    if (!groups.has(base)) groups.set(base, []);
    groups.get(base)!.push(cat);
  }

  const result = new Map<string, string[]>();

  for (const [base, variants] of groups) {
    if (variants.length >= 2 && base.length >= 4) {
      // This has seat-level variants - map each variant to ALL variants
      for (const v of variants) {
        result.set(v, variants);
      }
    } else {
      // Single variant - no expansion
      for (const v of variants) {
        result.set(v, [v]);
      }
    }
  }

  return result;
}

async function getCategoryVariants(category: string): Promise<string[]> {
  if (!categoryVariantsCache) {
    categoryVariantsCache = buildCategoryVariants();
  }
  const map = await categoryVariantsCache;
  return map.get(category) || [category];
}

/**
 * Calculate prediction score based on user percentile vs cutoff data.
 *
 * The CSV data stores a single rank/percentile per row, so opening === closing
 * for every record. We use proximity-based scoring:
 * - If >= closing percentile => Safe (you beat the cutoff, score 80-100)
 * - If within ~2% below => Moderate (close, might get in, score 60-80)
 * - If within ~5% below => Dream (worth a shot, score 25-50)
 * - Otherwise => No Chance (score 0-15)
 */
function calculateScore(
  percentile: number,
  closingPercentile: number,
  _openingPercentile: number
): number {
  const gap = closingPercentile - percentile;

  if (gap <= 0) {
    // At or above cutoff => Safe
    // Higher score when cutoff is closest to user's percentile (most relevant)
    // gap is negative here, so closing - percentile < 0 means user is above
    // e.g., gap=-0.4 => score=99.6, gap=-80 => score=80
    return Math.max(80, 100 + gap);
  } else if (gap <= 2) {
    // Within 2% below cutoff => Moderate
    return 60 + (1 - gap / 2) * 20;
  } else if (gap <= 5) {
    // Within 2-5% below cutoff => Dream
    return 25 + (1 - (gap - 2) / 3) * 25;
  } else {
    // More than 5% below cutoff => No Chance
    return Math.max(0, 15 - (gap - 5) * 3);
  }
}

function classifyLevel(score: number): "dream" | "moderate" | "safe" | "no-chance" {
  if (score >= 75) return "safe";
  if (score >= 50) return "moderate";
  if (score >= 20) return "dream";
  return "no-chance";
}

export async function predictColleges(
  input: PredictionInput
): Promise<{
  safe: PredictionResult[];
  moderate: PredictionResult[];
  dream: PredictionResult[];
  noChance: PredictionResult[];
  metadata: {
    totalPredictions: number;
    inputPercentile: number;
    inputCategory: string;
    inputGender: string | null;
    year: number;
    capRound: number;
    availableYears?: number[];
  };
}> {
  const year = input.year || 2024;
  const capRound = input.capRound || 0;  // 0 means all rounds
  const category = input.category || "GOPENS";

  // Build category filter: expand selected category to all seat-level variants
  // e.g., "GOPENS" → ["GOPENS", "GOPENH", "GOPENO"]
  const categoriesToMatch = await getCategoryVariants(category);

  // If user is Female, also include the Ladies variant (e.g., "GOPENS" -> also include "LOPENS")
  if (input.gender === "Female" && category.startsWith("G")) {
    const ladiesCategory = "L" + category.slice(1);
    if (ladiesCategory !== category) {
      const ladiesVariants = await getCategoryVariants(ladiesCategory);
      for (const v of ladiesVariants) {
        if (!categoriesToMatch.includes(v)) {
          categoriesToMatch.push(v);
        }
      }
    }
  }

  // If TFWS is checked, also include TFWS category
  if (input.tfws) {
    const tfwsVariants = await getCategoryVariants("TFWS");
    for (const v of tfwsVariants) {
      if (!categoriesToMatch.includes(v)) {
        categoriesToMatch.push(v);
      }
    }
  }

  // If minority is checked, also include MI (Minority Institute) category
  if (input.minority) {
    if (!categoriesToMatch.includes("MI")) {
      categoriesToMatch.push("MI");
    }
  }

  // Build the where clause - if capRound is 0, fetch all rounds
  const whereClause: any = {
    year,
    category: {
      in: categoriesToMatch,
    },
  };
  if (capRound > 0) {
    whereClause.capRound = capRound;
  }

  const cutoffRows = await prisma.cutoff.findMany({
    where: whereClause,
    include: {
      college: true,
      branch: true,
    },
    orderBy: {
      closingPercentile: "desc",
    },
  });

  const results: PredictionResult[] = [];
  const seen = new Set<string>();

  for (const row of cutoffRows) {
    // Deduplicate: same college + branch + category may have multiple stages (I, II)
    // Keep only the first occurrence (highest closingPercentile due to ORDER BY)
    const key = `${row.collegeId}-${row.branchId}-${row.category}`;
    if (seen.has(key)) continue;
    seen.add(key);

    const score = calculateScore(
      input.percentile,
      row.closingPercentile,
      row.openingPercentile
    );
    const level = classifyLevel(score);

    // Apply filters
    if (
      input.preferredBranches &&
      input.preferredBranches.length > 0 &&
      !input.preferredBranches.some(
        (b) =>
          row.branch.name.toLowerCase().includes(b.toLowerCase()) ||
          row.branch.code.toLowerCase().includes(b.toLowerCase())
      )
    ) {
      continue;
    }

    // Apply city filter with suburb expansion
    // If user selected "Navi Mumbai", also match colleges in "Panvel", "Vashi", etc.
    if (input.city) {
      const userCity = input.city.toLowerCase();
      const collegeCity = row.college.city.toLowerCase();
      
      // Direct match or college city is a known suburb of selected city
      const isMatchingCity =
        collegeCity === userCity ||
        SUBURB_TO_PARENT[collegeCity] === userCity ||
        // Also check if college city is listed as a suburb of the selected city
        (CITY_SUBURBS[input.city] &&
          CITY_SUBURBS[input.city].some(
            (s) => s.toLowerCase() === collegeCity
          ));
      
      if (!isMatchingCity) {
        continue;
      }
    }

    results.push({
      collegeId: row.collegeId,
      collegeName: row.college.name,
      collegeSlug: row.college.slug,
      branchId: row.branchId,
      branchName: row.branch.name,
      category: row.category,
      closingPercentile: row.closingPercentile,
      openingPercentile: row.openingPercentile,
      closingRank: row.closingRank,
      openingRank: row.openingRank,
      year: row.year,
      capRound: row.capRound,
      level,
      score,
    });
  }

  results.sort((a, b) => b.score - a.score);

  // Limit top results
  const topResults = results.slice(0, 500);

  const safe = topResults.filter((r) => r.level === "safe");
  const moderate = topResults
    .filter((r) => r.level === "moderate");
  const dream = topResults.filter((r) => r.level === "dream");
  const noChance = topResults
    .filter((r) => r.level === "no-chance");

  // If no results, check which years have data for this category
  // so the UI can show a helpful message
  let availableYears: number[] = [];
  if (results.length === 0) {
    // Fetch available years (avoid Prisma distinct issues)
    const yearRows = await prisma.cutoff.findMany({
      where: {
        category: { in: categoriesToMatch },
      },
      select: { year: true },
    });
    const yearSet = new Set(yearRows.map((r) => r.year));
    availableYears = [...yearSet].sort((a, b) => a - b);
  }

  return {
    safe,
    moderate,
    dream,
    noChance,
    metadata: {
      totalPredictions: results.length,
      inputPercentile: input.percentile,
      inputCategory: category,
      inputGender: input.gender || null,
      year,
      capRound,
      availableYears,
    },
  };
}

export async function compareColleges(
  collegeIds: string[]
): Promise<
  {
    college: { id: string; name: string; slug: string; city: string; fees: number; naacGrade: string; nirfRank: number };
    cutoffs: {
      year: number;
      capRound: number;
      category: string;
      branchName: string;
      closingPercentile: number;
      openingPercentile: number;
    }[];
  }[]
> {
  const colleges = await prisma.college.findMany({
    where: { id: { in: collegeIds } },
    include: {
      cutoffs: {
        include: { branch: true },
        orderBy: [{ year: "desc" }, { capRound: "desc" }],
        take: 30,
      },
    },
  });

  return colleges.map((c) => ({
    college: {
      id: c.id,
      name: c.name,
      slug: c.slug,
      city: c.city,
      fees: c.fees,
      naacGrade: c.naacGrade,
      nirfRank: c.nirfRank,
    },
    cutoffs: c.cutoffs.map((cut) => ({
      year: cut.year,
      capRound: cut.capRound,
      category: cut.category,
      branchName: cut.branch.name,
      closingPercentile: cut.closingPercentile,
      openingPercentile: cut.openingPercentile,
    })),
  }));
}

File: cet_app/lib/utils/prediction_engine.dart
import '../models/prediction_model.dart';
import '../models/college_model.dart';
import '../models/cutoff_model.dart';
import '../data/data_service.dart';

/// Prediction engine ported from the TypeScript version.
/// Predicts colleges based on user percentile and filters.
class PredictionEngine {
  final DataService _data = DataService();

  /// Map from base category to all its seat-level variants (S, H, O)
  static final Map<String, List<String>> _categoryVariants = {
    'GOPEN': ['GOPENS', 'GOPENH', 'GOPENO'],
    'GSC': ['GSCS', 'GSCH', 'GSCO'],
    'GST': ['GSTS', 'GSTH', 'GSTO'],
    'GVJ': ['GVJS', 'GVJH', 'GVJO'],
    'GNT1': ['GNT1S', 'GNT1H', 'GNT1O'],
    'GNT2': ['GNT2S', 'GNT2H', 'GNT2O'],
    'GNT3': ['GNT3S', 'GNT3H', 'GNT3O'],
    'GOBC': ['GOBCS', 'GOBCH', 'GOBCO'],
    'LOPEN': ['LOPENS', 'LOPENH', 'LOPENO'],
    'LSC': ['LSCS', 'LSCH', 'LSCO'],
    'LST': ['LSTS', 'LSTH', 'LSTO'],
    'LVJ': ['LVJS', 'LVJH', 'LVJO'],
    'LNT1': ['LNT1S', 'LNT1H', 'LNT1O'],
    'LNT2': ['LNT2S', 'LNT2H', 'LNT2O'],
    'LOBC': ['LOBCS', 'LOBCH', 'LOBCO'],
  };

  /// Get category variants for a given category code
  List<String> getCategoryVariants(String category) {
    // Check for TFWS, EWS, MI, DEFOPENS, etc.
    if (category == 'TFWS' || category == 'EWS' || category == 'MI') {
      return [category];
    }

    // Check for DEF categories
    if (category.startsWith('DEF') || category.startsWith('PWD') || category.startsWith('SDEF')) {
      return [category];
    }

    // Strip trailing S/H/O
    for (final entry in _categoryVariants.entries) {
      if (category.startsWith(entry.key)) {
        final lastChar = category.substring(entry.key.length);
        if (lastChar == 'S' || lastChar == 'H' || lastChar == 'O') {
          return _categoryVariants[entry.key]!;
        }
        // Also include if it matches exactly (e.g., "GOPENS" input)
        if (entry.value.contains(category)) {
          return entry.value;
        }
      }
    }

    // If it's a base category without suffix, try to find it
    if (_categoryVariants.containsKey(category)) {
      return _categoryVariants[category]!;
    }

    return [category];
  }

  /// City suburb expansion (same as TypeScript version)
  static const Map<String, List<String>> citySuburbs = {
    'Navi Mumbai': ['Panvel', 'Vashi', 'Nerul', 'Airoli', 'Ghansoli', 'Belapur', 'Kharghar', 'Kamothe', 'Kalamboli', 'New Panvel'],
    'Mumbai': ['Andheri', 'Bhayander', 'Bhiwandi', 'Kandivali', 'Matunga', 'Boisar', 'Palghar'],
    'Pune': ['Pimpri', 'Chinchwad', 'Haveli', 'Pisoli', 'Ravet', 'Sasewadi', 'Talegaon', 'Wagholi', 'Avasari Khurd'],
    'Sangli': ['Miraj'],
    'Amravati': ['Badnera', 'Shegaon'],
    'Nashik': ['Nepti', 'Nashik', 'Nadurbar'],
    'Kalyan': ['Dombivli', 'Ulhasnagar'],
    'Kolhapur': ['Ichalkaranji', 'Panhala'],
    'Solapur': ['Barshi'],
    'Nagpur': ['Ramtek', 'Wardha'],
  };

  // Build reverse lookup
  static final Map<String, String> _suburbToParent = () {
    final map = <String, String>{};
    for (final entry in citySuburbs.entries) {
      for (final suburb in entry.value) {
        map[suburb.toLowerCase()] = entry.key.toLowerCase();
      }
    }
    return map;
  }();

  /// Calculate prediction score (0-100) based on user percentile vs cutoff
  double _calculateScore(double percentile, double closingPercentile) {
    final gap = closingPercentile - percentile;

    if (gap <= 0) {
      // At or above cutoff => Safe
      return (100 + gap).clamp(80, 100);
    } else if (gap <= 2) {
      // Within 2% below cutoff => Moderate
      return 60 + (1 - gap / 2) * 20;
    } else if (gap <= 5) {
      // Within 2-5% below cutoff => Dream
      return 25 + (1 - (gap - 2) / 3) * 25;
    } else {
      // More than 5% below cutoff => No Chance
      return (15 - (gap - 5) * 3).clamp(0, 15);
    }
  }

  /// Classify level based on score
  PredictionLevel _classifyLevel(double score) {
    if (score >= 75) return PredictionLevel.safe;
    if (score >= 50) return PredictionLevel.moderate;
    if (score >= 20) return PredictionLevel.dream;
    return PredictionLevel.noChance;
  }

  /// Predict colleges for the given input
  Predictions predict({
    required double percentile,
    required String category,
    String? gender,
    String? city,
    String? branch,
    bool tfws = false,
    bool minority = false,
    int year = 2024,
    int capRound = 0, // 0 = all rounds
  }) {
    // Expand category to variants
    List<String> categoriesToMatch = List.from(getCategoryVariants(category));

    // If Female, also include Ladies variants
    if (gender == 'Female' && category.startsWith('G')) {
      final ladiesCategory = 'L${category.substring(1)}';
      final ladiesVariants = getCategoryVariants(ladiesCategory);
      for (final v in ladiesVariants) {
        if (!categoriesToMatch.contains(v)) categoriesToMatch.add(v);
      }
    }

    // If TFWS, include TFWS category
    if (tfws) {
      if (!categoriesToMatch.contains('TFWS')) categoriesToMatch.add('TFWS');
    }

    // If minority, include MI category
    if (minority) {
      if (!categoriesToMatch.contains('MI')) categoriesToMatch.add('MI');
    }

    // Filter cutoffs
    var filtered = _data.cutoffs.where((c) {
      if (c.year != year) return false;
      if (capRound > 0 && c.capRound != capRound) return false;
      if (!categoriesToMatch.contains(c.category)) return false;
      return true;
    }).toList();

    // Sort by closing percentile descending
    filtered.sort((a, b) => b.percentile.compareTo(a.percentile));

    // Save branch filter string before it gets shadowed by the loop variable
    final branchFilter = branch;

    // Deduplicate by college + branch + category
    final seen = <String>{};
    final results = <PredictionResult>[];

    for (final c in filtered) {
      final key = '${c.collegeCode}-${c.branchCode}-${c.category}';
      if (seen.contains(key)) continue;
      seen.add(key);

      final college = _data.getCollege(c.collegeCode);
      final branchObj = _data.getBranch(c.branchCode);
      if (college == null || branchObj == null) continue;

      // Apply city filter with suburb expansion
      if (city != null && city.isNotEmpty) {
        final userCity = city.toLowerCase();
        final collegeCity = college.city.toLowerCase();

        // Check direct match or suburb match
        final parentCity = _suburbToParent[collegeCity];
        final isMatching = collegeCity == userCity ||
            parentCity == userCity ||
            (citySuburbs[city]?.any((s) => s.toLowerCase() == collegeCity) ?? false);

        if (!isMatching) continue;
      }

      // Apply branch filter
      if (branchFilter != null && branchFilter.isNotEmpty) {
        if (!branchObj.name.toLowerCase().contains(branchFilter.toLowerCase()) &&
            !branchObj.code.toLowerCase().contains(branchFilter.toLowerCase())) {
          continue;
        }
      }

      final score = _calculateScore(percentile, c.percentile);
      final level = _classifyLevel(score);

      results.add(PredictionResult(
        collegeCode: c.collegeCode,
        collegeName: college.name,
        collegeCity: college.city,
        branchCode: c.branchCode,
        branchName: branchObj.name,
        category: c.category,
        closingPercentile: c.percentile,
        closingRank: c.rank,
        year: c.year,
        capRound: c.capRound,
        level: level,
        score: score,
      ));
    }

    // Sort by score descending
    results.sort((a, b) => b.score.compareTo(a.score));

    // Limit to top 500
    final topResults = results.take(500).toList();

    // Check available years if no results
    List<int> availableYears = [];
    if (results.isEmpty) {
      final yearCategories = _data.cutoffs
          .where((c) => categoriesToMatch.contains(c.category))
          .map((c) => c.year)
          .toSet();
      availableYears = yearCategories.toList()..sort();
    }

    return Predictions(
      safe: topResults.where((r) => r.level == PredictionLevel.safe).toList(),
      moderate: topResults.where((r) => r.level == PredictionLevel.moderate).toList(),
      dream: topResults.where((r) => r.level == PredictionLevel.dream).toList(),
      noChance: topResults.where((r) => r.level == PredictionLevel.noChance).toList(),
      metadata: PredictionMetadata(
        totalPredictions: results.length,
        inputPercentile: percentile,
        inputCategory: category,
        inputGender: gender,
        year: year,
        capRound: capRound,
        availableYears: availableYears,
      ),
    );
  }
}

class Predictions {
  final List<PredictionResult> safe;
  final List<PredictionResult> moderate;
  final List<PredictionResult> dream;
  final List<PredictionResult> noChance;
  final PredictionMetadata metadata;

  const Predictions({
    required this.safe,
    required this.moderate,
    required this.dream,
    required this.noChance,
    required this.metadata,
  });

  bool get isEmpty => safe.isEmpty && moderate.isEmpty && dream.isEmpty && noChance.isEmpty;
}
